receiver/prometheusreceiver/metrics_receiver.go

```diff
@@ -92,36 +110,51 @@ func (r *pReceiver) Start(_ context.Context, host component.Host) error {
 
 	allocConf := r.cfg.TargetAllocator
 	if allocConf != nil {
-		go func() {
-			// immediately sync jobs and not wait for the first tick
-			savedHash, _ := r.syncTargetAllocator(uint64(0), allocConf, baseCfg)
-			r.targetAllocatorIntervalTicker = time.NewTicker(allocConf.Interval)
-			for {
-				<-r.targetAllocatorIntervalTicker.C
-				hash, err := r.syncTargetAllocator(savedHash, allocConf, baseCfg)
-				if err != nil {
-					r.settings.Logger.Error(err.Error())
-					continue
-				}
-				savedHash = hash
-			}
-		}()
+		err = r.initTargetAllocator(allocConf, baseCfg)
+		if err != nil {
+			return err
+		}
 	}
 
+	r.reloadReady.Close()
+
 	return nil
 }
 
+func (r *pReceiver) initTargetAllocator(allocConf *targetAllocator, baseCfg *config.Config) error {
+	r.settings.Logger.Info("Starting target allocator discovery")
+	// immediately sync jobs and not wait for the first tick
+	savedHash, err := r.syncTargetAllocator(uint64(0), allocConf, baseCfg)
+	if err != nil {
+		return err
+	}
+	r.targetAllocatorIntervalTicker = time.NewTicker(allocConf.Interval)
+	go func() {
```

Comment:
Wondering why the decision to pull this into its own function and move the `go func() {` block down after the call to `r.syncTargetAllocator`?

Reply:
That's to pull jobs before the scrape manager has a chance to begin so that as soon as the collector's scraping begins, it already has targets from the target allocator. It also means we'll fail fast if the target allocator isn't up and healthy.